gravitee-exchange-controller/gravitee-exchange-controller-core/src/main/java/io/gravitee/exchange/controller/core/channel/primary/PrimaryChannelManager.java

```diff
@@ -141,13 +140,25 @@ private void electPrimaryChannel(final String targetId) {
             return;
         }
         if (!channelIds.contains(previousPrimaryChannelId)) {
-            //noinspection deprecation
-            @SuppressWarnings("java:S1874")
-            String newPrimaryChannelId = channelIds.get(RandomUtils.nextInt(0, channelIds.size()));
-            primaryChannelCache.put(targetId, newPrimaryChannelId);
-            primaryChannelElectedEventTopic.publish(
-                PrimaryChannelElectedEvent.builder().targetId(targetId).channelId(newPrimaryChannelId).build()
-            );
+            String newPrimaryChannelId = getRandomChannel(channelIds);
+            if (newPrimaryChannelId != null) {
+                primaryChannelCache.put(targetId, newPrimaryChannelId);
+                primaryChannelElectedEventTopic.publish(
+                    PrimaryChannelElectedEvent.builder().targetId(targetId).channelId(newPrimaryChannelId).build()
+                );
+            }
+        }
+    }
+
+    private String getRandomChannel(Set<String> channelIds) {
+        int randomIndex = ThreadLocalRandom.current().nextInt(channelIds.size());
+        int i = 0;
+        for (String channelId : channelIds) {
+            if (i == randomIndex) {
+                return channelId;
+            }
+            i++;
         }
+        return null;
```

Comment:
How is that even possible ?

Reply:
if the channels list changed during the loop for example